pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/ContextImpl.java

```diff
@@ -706,4 +714,67 @@ public void close() {
             logger.warn("Failed to close producers", e);
         }
     }
+
+    @Override
+    public void seek(String topic, int partition, MessageId messageId) throws PulsarClientException {
+        Consumer<?> consumer = getConsumer(topic, partition);
+        consumer.seek(messageId);
+    }
+
+    @Override
+    public void pause(String topic, int partition) throws PulsarClientException {
+        getConsumer(topic, partition).pause();
+    }
+
+    @Override
+    public void resume(String topic, int partition) throws PulsarClientException {
+        getConsumer(topic, partition).resume();
+    }
+
+    public void setInputConsumers(List<Consumer<?>> inputConsumers) {
+        this.inputConsumers = inputConsumers;
+        inputConsumers.stream()
+            .flatMap(consumer ->
+                    consumer instanceof MultiTopicsConsumerImpl
+                            ? ((MultiTopicsConsumerImpl<?>) consumer).getConsumers().stream()
+                            : Stream.of(consumer))
+            .forEach(consumer -> topicConsumers.putIfAbsent(TopicName.get(consumer.getTopic()), consumer));
+    }
+
+    @VisibleForTesting
+    Consumer<?> getConsumer(String topic, int partition) throws PulsarClientException {
+        if (inputConsumers == null) {
+            throw new PulsarClientException("Getting consumer is not supported");
+        }
+        for (int i = 0; i < 2; i++) {
```

Comment:
Why is this for loop needed?

Reply:
Two attempts to get the consumer:
1. Try to get the consumer. 
2. If not found, reprocess MultiTopicsConsumers in case new consumers appeared (happens on repartition or a new topic that matches provided pattern if the pattern is used), 
3. give it another try.

Reply:
This doesn't seem like the right way to do this.  Retrying twice seems arbitrary.  I would rather let the user implement how many retries they want.

Reply:
@jerrypeng this is not arbitrary.
I have no access to callbacks/events when the MultiTopicsConsumers adds another partition or a topic (for pattern-based subscription). MultiTopicsConsumers do have that internally.
So if the check of previously cached topic/partition -> consumer map finds nothing it means that it:
- either doesn't exist
- or the topic/partition (and consumer) got added after the initial map was built.
In that case it makes sense to check MultiTopicsConsumers once to update the map and try again. 
I don't think blocking and polling is justified, at least I don't have such usecase.

Reply:
@dlg99 if partitions / topics are added dynamically because of topic pattern match or increases in partition count, how would the function know when that happened and why would trying to the consumers twice help with that?  There is an inherent race condition there.  If can always argue there can be a scenario that topics/partitions got created after you called this method.  If the goal of the method is to get new consumers for new topics / partitions then this API is not appropriate.

Reply:
We talked offline about reasons it is needed.
I expressed the retry more clearly and added comments. Please take a look.